File: Sources/Golos/VoiceMessage.swift
import AppKit
import Foundation

/// Голосовое сообщение: запись сжимается и кладётся в буфер файлом.
/// Дальше её вставляют в чат обычным Cmd+V.
///
/// Формат ogg с opus, а не m4a: мессенджеры показывают голосовым сообщением
/// именно его, остальное приезжает вложением с плеером. 48 кГц моно, 24 кбит,
/// профиль voip — то же, что пишут сами мессенджеры. Выходит около 178 КБ
/// на минуту против 1875 у исходного WAV.
enum VoiceMessage {
    enum VoiceError: LocalizedError {
        case noEncoder
        case encodingFailed(String)

        var errorDescription: String? {
            switch self {
            case .noEncoder:
                return "Сжатие звука недоступно: в приложении нет ffmpeg"
            case .encodingFailed(let detail):
                return "Не удалось сжать запись: \(detail)"
            }
        }
    }

    /// Готовые файлы живут рядом с историей, а не во временной папке системы:
    /// в буфере лежит ссылка, и файл обязан пережить вставку.
    static var directory: URL {
        Config.directory.appendingPathComponent("voice")
    }

    /// Сколько держим отправленные файлы. Ссылка в буфере живёт, пока живёт
    /// файл, поэтому убирать сразу нельзя.
    private static let retentionHours = 24.0

    /// Сжимает запись и кладёт её в буфер. Возвращает путь к файлу.
    static func copyToClipboard(wav: Data, completion: @escaping (Result<URL, Error>) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            let result = Result { try encode(wav: wav) }
            DispatchQueue.main.async {
                switch result {
                case .failure(let error):
                    completion(.failure(error))
                case .success(let url):
                    let board = NSPasteboard.general
                    board.clearContents()
                    board.writeObjects([url as NSURL])
                    purge()
                    completion(.success(url))
                }
            }
        }
    }

    private static func encode(wav: Data) throws -> URL {
        guard let ffmpeg = MediaDecoder.ffmpegPath() else { throw VoiceError.noEncoder }

        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let stamp = ISO8601DateFormatter.filenameSafe.string(from: Date())
        let source = directory.appendingPathComponent("\(stamp).wav")
        let target = directory.appendingPathComponent("voice-\(stamp).ogg")
        try wav.write(to: source)
        defer { try? FileManager.default.removeItem(at: source) }

        let process = Process()
        process.executableURL = URL(fileURLWithPath: ffmpeg)
        process.arguments = ["-y", "-loglevel", "error", "-i", source.path,
                             "-c:a", "libopus", "-b:a", "24k", "-ar", "48000",
                             "-ac", "1", "-application", "voip",
                             "-f", "ogg", target.path]
        process.environment = Whisper.childEnvironment
        let errors = Pipe()
        process.standardOutput = FileHandle.nullDevice
        process.standardError = errors
        try process.run()
        let complaint = errors.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()

        guard process.terminationStatus == 0,
              let size = try? FileManager.default.attributesOfItem(atPath: target.path)[.size] as? Int,
              size > 0
        else {
            let detail = String(data: complaint, encoding: .utf8)?
                .trimmingCharacters(in: .whitespacesAndNewlines) ?? "код \(process.terminationStatus)"
            throw VoiceError.encodingFailed(detail.isEmpty ? "код \(process.terminationStatus)" : detail)
        }
        return target
    }

    /// Чистит вчерашние файлы. Свежий не трогаем никогда: на него смотрит буфер.
    private static func purge() {
        let files = (try? FileManager.default.contentsOfDirectory(
            at: directory, includingPropertiesForKeys: [.contentModificationDateKey])) ?? []
        let deadline = Date().addingTimeInterval(-retentionHours * 3600)
        var removed = 0
        for file in files {
            let modified = (try? file.resourceValues(forKeys: [.contentModificationDateKey]))?
                .contentModificationDate ?? Date()
            if modified < deadline, (try? FileManager.default.removeItem(at: file)) != nil { removed += 1 }
        }
        if removed > 0 { Log.write("голосовые: удалено файлов \(removed)") }
    }
}

File: windows/src/VoiceMessage.cs
using System.Diagnostics;

namespace Golos;

/// <summary>
/// Голосовое сообщение: запись сжимается и кладётся в буфер файлом.
/// Дальше её вставляют в чат обычным Ctrl+V.
///
/// Формат ogg с opus, а не m4a: мессенджеры показывают голосовым сообщением
/// именно его, остальное приезжает вложением с плеером. 48 кГц моно, 24 кбит,
/// профиль voip — то же, что пишут сами мессенджеры. Выходит около 178 КБ
/// на минуту против 1875 у исходного WAV.
/// </summary>
public static class VoiceMessage
{
    /// <summary>
    /// Готовые файлы живут рядом с настройками, а не во временной папке:
    /// в буфере лежит ссылка, и файл обязан пережить вставку.
    /// </summary>
    public static string Directory => Path.Combine(Config.Directory, "voice");

    /// <summary>Сколько держим отправленные файлы.</summary>
    private const double RetentionHours = 24;

    /// <summary>Сжимает запись и возвращает путь к файлу либо текст ошибки.</summary>
    public static (string? path, string? error) Encode(byte[] wav)
    {
        var ffmpeg = MediaDecoder.FindFfmpegPublic();
        if (ffmpeg == null) return (null, "Сжатие звука недоступно: рядом нет ffmpeg");

        try
        {
            System.IO.Directory.CreateDirectory(Directory);
            var stamp = DateTime.Now.ToString("yyyy-MM-dd'T'HHmmss");
            var source = Path.Combine(Directory, $"{stamp}.wav");
            var target = Path.Combine(Directory, $"voice-{stamp}.ogg");
            File.WriteAllBytes(source, wav);

            try
            {
                var info = new ProcessStartInfo
                {
                    FileName = ffmpeg,
                    UseShellExecute = false,
                    CreateNoWindow = true,
                    RedirectStandardError = true
                };
                foreach (var a in new[] { "-y", "-loglevel", "error", "-i", source,
                                          "-c:a", "libopus", "-b:a", "24k", "-ar", "48000",
                                          "-ac", "1", "-application", "voip",
                                          "-f", "ogg", target })
                    info.ArgumentList.Add(a);

                using var process = Process.Start(info);
                if (process == null) return (null, "ffmpeg не запустился");
                var complaint = process.StandardError.ReadToEnd().Trim();
                process.WaitForExit();

                if (process.ExitCode != 0 || !File.Exists(target) || new FileInfo(target).Length == 0)
                    return (null, complaint.Length > 0 ? complaint : $"ffmpeg вышел с кодом {process.ExitCode}");

                Purge();
                return (target, null);
            }
            finally
            {
                try { File.Delete(source); } catch { }
            }
        }
        catch (Exception e) { return (null, e.Message); }
    }

    /// <summary>Кладёт готовый файл в буфер обмена ссылкой.</summary>
    public static void CopyToClipboard(string path)
    {
        var files = new System.Collections.Specialized.StringCollection { path };
        Clipboard.SetFileDropList(files);
    }

    /// <summary>Чистит вчерашние файлы. Свежий не трогаем: на него смотрит буфер.</summary>
    private static void Purge()
    {
        try
        {
            var deadline = DateTime.Now.AddHours(-RetentionHours);
            var removed = 0;
            foreach (var file in System.IO.Directory.GetFiles(Directory))
            {
                if (File.GetLastWriteTime(file) >= deadline) continue;
                try { File.Delete(file); removed++; } catch { }
            }
            if (removed > 0) Log.Write($"голосовые: удалено файлов {removed}");
        }
        catch { }
    }
}
